### DISCUSSION TITLE
How to disable standard authentication for SSO users

### DISCUSSION THREAD

**Discussion by drbng** [keycloak/keycloak#24229] [Q&A]
Hello,

I would like to know how can I disabled the standard authentication (login/password) for all users that are linked to an Identity Provider 

example: 

In the filed Identity provider links: 

![image](https://github.com/keycloak/keycloak/assets/115652595/a8d450b4-928d-44e1-81f1-e65fff2c3137)

These kind of users have to be authenticated through Single Sign On method and I wish to disable the login/Password authentication on my keycloak 21 version. With the old keycloak version, I was allowed to disable the credential method example:

![image](https://github.com/keycloak/keycloak/assets/115652595/cc6b65f2-a020-4633-bf22-646741a32932)

With the new version I can't see the disable option example:

![image](https://github.com/keycloak/keycloak/assets/115652595/1061277c-8fa1-46c0-a99a-2492ba376098)

Thank you,

**Comment by LucasVanHaaren** [comment]
Hello everyone,

I'm trying to achieve the same thing, ie: disable local login for users who have an identity linked to their account but continue to allow local users (who have no identity linked) to login with a username/password.

Is there a simple way to do it for now ?

**Comment by thomasdarimont** [comment]
One way to achive this is to use an "Identifer First" Login Flow, where you ask your users for their username/email in a first step.
With the given username/email you perform a lookup in your user base and detect iff the user exists. If the user was not found you could show the password form. If the  user was found and has an IdP link attached, you could apply some logic to redirect the user to his / her home IdP automatically. Otherwise you show the password form for the user to allow the user to perform a local login.

If you want to support multiple Identity providers, you need a mechanism to determine the "home" realm for a user or present an narrowed down IdP selection. The home IdP could be derived from metadata added to the IdP, e.g. an regex of allowed email domains.

Some extension examples which might help you:
- [dynamicidp forward the known user to the (single) target IdP](https://github.com/thomasdarimont/keycloak-project-example/tree/main/keycloak/extensions/src/main/java/com/github/thomasdarimont/keycloak/custom/auth/dynamicidp)
- [IdP based on Username, e.g. email domain](https://github.com/thomasdarimont/keycloak-project-example/blob/main/keycloak/extensions/src/main/java/com/github/thomasdarimont/keycloak/custom/auth/idpselection/AcmeDynamicIdpLookupUsernameForm.java)
- [home-realm-discovery](https://github.com/sventorben/keycloak-home-idp-discovery)

**Reply by LucasVanHaaren** [reply]
Thank you, I will take a look ASAP !

**Comment by LucasVanHaaren** [comment]
Hi @drbng, I finally managed to realize this, here’s how I did it: 

- I created a realm_role named "brokered_identity", it will be assigned to each user who linked his account to an external identity 
- I created an Identity Provider Mapper that assigns this role :
![image](https://github.com/keycloak/keycloak/assets/29121316/7d52d457-3cd8-4739-961c-ec75f189bb77)

- I modified the authentication flow by browser to check that a user connecting via the login/password form does not have this role :
![image](https://github.com/keycloak/keycloak/assets/29121316/dec83de9-1cd4-4572-acda-92202b437a94)

---

Note that I also tried the [keycloak-home-idp-discovery](https://github.com/sventorben/keycloak-home-idp-discovery) plugin (which is quite great) as @thomasdarimont suggested, but I will not use it as the solution above suits well for my use case and I don't want to bother with third-party extensions.

**Reply by mvanhulsentop** [reply]
With this approach, isn't it still possible for a user to unlink the IdP from the account (via the account management), and set a password. That way, a user can 'convert' the IdP based account to a local KeyCloak account.

**Reply by LucasVanHaaren** [reply]
Yes, a user can unlink an identity from his account via the console, and then use a username/password couple.
The only solution I know to not allow this is to deactivate the account-console client in the realm (which is quite brutal).

**Comment by drbng** [comment]
Hello,

Thank you very much, it is really helpful !
I will do the same

**Comment by ryan-morris** [comment]
What seems to have been an acceptable work around here by blocking sign in with a password still leaves some edge case issues for me. I put more information on https://github.com/keycloak/keycloak/discussions/27152 but tldr is that letting the user set a password can override the automatic redirect specified on an organization idp, and without allowing the idp to be shown on the login page, the user has no way to login.
